fuel-bft/src/metadata.rs

```diff
@@ -0,0 +1,210 @@
+        let key = self.id(&round)?;
+
+        let message =
+            <Self::Network as Network>::Message::signed(signer, key, round, state, block)?;
+
+        Ok(message)
+    }
+
+    /// Fetch the public key of the node for a given round
+    fn public_key(&self, round: &HeightRound) -> Result<Borrown<'_, PublicKey>, Self::Error> {
+        let id = self.id(round)?;
+        let public = self.signer().id_public(id)?;
+
+        Ok(public)
+    }
+
+    /// Number of validators for a given height
+    fn validators_count(&self, height: u64) -> Result<usize, Self::Error> {
+        self.filter_round(height).map(|i| i.count())
+    }
+
+    /// Evaluate the state count for a given round, including the validators that are in subsequent
+    /// states.
+    fn evaluate_state_count(&self, round: &HeightRound, state: State) -> usize {
+        let current = self.state_count(round, state);
+        let subsequent: usize = state.map(|s| self.state_count(round, s)).sum::<usize>();
+
+        current + subsequent
+    }
+
+    /// Upgrade a validator state, returning true if there was a change
+    fn upgrade_validator_state(&mut self, round_key: &RoundValidator, state: State) -> bool {
+        match self.validator_state(round_key) {
+            None => {
+                self.set_validator_state(round_key, state);
+
+                #[cfg(feature = "trace")]
+                tracing::debug!(
+                    "upgrading state; validator: {:08x}, round: {}, state: {:?}",
+                    round_key.validator(),
+                    round_key.round(),
+                    state
+                );
+
+                true
+            }
+
+            Some(s) if state > s => {
+                self.set_validator_state(round_key, state);
+
+                #[cfg(feature = "trace")]
+                tracing::debug!(
+                    "upgrading state; validator: {:08x}, round: {}, state: {:?}",
+                    round_key.validator(),
+                    round_key.round(),
+                    state
+                );
+
+                true
+            }
+
+            _ => {
+                #[cfg(feature = "trace")]
+                tracing::trace!(
+                    "upgrading state skipped; validator: {:08x}, round: {}, state: {:?}",
+                    round_key.validator(),
+                    round_key.round(),
+                    state
+                );
+
+                false
+            }
+        }
+    }
+
+    /// Compute the round leader.
+    fn leader(&self, round: &HeightRound) -> Result<Self::PublicKey, Self::Error> {
+        #[cfg(feature = "trace")]
+        tracing::trace!("choosing leader for round {}", round);
+
+        let count = self.validators_count(round.height())? as u64;
+
+        if count == 0 {
+            return Err(Error::ValidatorNotFound.into());
+        }
+
+        let h = round.height();
+        let r = round.round();
+
+        let index = (h + r) % count;
```

Comment:
Now that there is some tracking for the total round number, could we use something like `(self.last_commit_total_rounds()) % count` instead?

Reply:
Yes, that would be a good change. However, the current leader must take into account also the spared time since the last commit, so it can evaluate what is the current leader properly. I need to give a careful thought about this one